torch/csrc/jit/ir/alias_analysis.cpp

```diff
@@ -90,7 +91,7 @@ bool isMutableTypeImpl(
   // getMutableTypePtrImpl
   auto kind = type->kind();
   if (kind == TypeKind::TensorType || kind == TypeKind::ListType ||
-      kind == TypeKind::ClassType || kind == TypeKind::DictType) {
+      kind == TypeKind::ClassType || kind == TypeKind::DictType || kind == TypeKind::UnionType) {
```

Comment:
We shouldn't have Union here for the same reason we don't have Optional here. If it is a union of immutable types than the type itself is also immutable.